cilium/cilium
# Refactor /endpoint/{id} for API 1.0 stability

*Note: This PR is a diff to #3431. Please don't merge.
See #3280 for background

Summary of changes:
- Refactor the endpoint data we present in the API. This reflects the Spec/Status scheme, groups subfields, and renames some fields and types.
- The CEP controller is more resilient to invalid objects (which happen when we change the schema).

How to test (optional):
all tests should pass.

```go
type Endpoint struct {

        // The cilium-agent-local ID of the endpoint
        ID int64 `json:"id,omitempty"`

        // The desired configuration state of the endpoint
        Spec *EndpointConfigurationSpec `json:"spec,omitempty"`

        // The desired and realized configuration state of the endpoint
        Status *EndpointStatus `json:"status,omitempty"`
}

type EndpointStatus struct {

        // Status of internal controllers attached to this endpoint
        Controllers ControllerStatuses `json:"controllers"`

        // Unique identifiers for this endpoint from outside cilium
        ExternalIdentifiers *EndpointIdentifiers `json:"externalIdentifiers,omitempty"`

        // Summary overall endpoint & subcomponent health
        Health *EndpointHealth `json:"health,omitempty"`

        // The security identity for this endpoint
        Identity *Identity `json:"identity,omitempty"`

        // Labels applied to this endpoint
        Labels *LabelConfigurationStatus `json:"labels,omitempty"`

        // Most recent status log. See endpoint/{id}/log for the complete log.
        Log EndpointStatusLog `json:"log"`

        // Networking properties of the endpoint
        Networking *EndpointNetworking `json:"networking,omitempty"`

        // The policy applied to this endpoint from the policy repository
        Policy *EndpointPolicyStatus `json:"policy,omitempty"`

        // The configuration in effect on this endpoint
        Realized *EndpointConfigurationSpec `json:"realized,omitempty"`

        // Current state of endpoint
        // Required: true
        State EndpointState `json:"state"`
}

type EndpointPolicyStatus struct {
        // The policy in the datapath for this endpoint
        Realized *EndpointPolicy `json:"realized,omitempty"`

        // The policy that should apply to this endpoint
        Spec *EndpointPolicy `json:"spec,omitempty"`

        // The policy revision currently enforced in the proxy for this endpoint
        ProxyPolicyRevision int64 `json:"proxy-policy-revision,omitempty"`

        // Statistics of the proxy redirects configured for this endpoint
        ProxyStatistics []*ProxyStatistics `json:"proxy-statistics"`
}

type EndpointPolicy struct {

        // List of identities to which this endpoint is allowed to communicate
        //
        AllowedEgressIdentities []int64 `json:"allowed-egress-identities"`

        // List of identities allowed to communicate to this endpoint
        //
        AllowedIngressIdentities []int64 `json:"allowed-ingress-identities"`

        // Build number of calculated policy in use
        Build int64 `json:"build,omitempty"`

        // cidr policy
        CidrPolicy *CIDRPolicy `json:"cidr-policy,omitempty"`

        // Own identity of endpoint
        ID int64 `json:"id,omitempty"`

        // l4
        L4 *L4Policy `json:"l4,omitempty"`

        // Whether policy enforcement is enabled (ingress, egress, both or none)
        PolicyEnabled EndpointPolicyEnabled `json:"policy-enabled,omitempty"`

        // The agent-local policy revision
        PolicyRevision int64 `json:"policy-revision,omitempty"`
}

type EndpointIdentifiers struct {

        // ID assigned by container runtime
        ContainerID string `json:"container-id,omitempty"`

        // Name assigned to container
        ContainerName string `json:"container-name,omitempty"`

        // Docker endpoint ID
        DockerEndpointID string `json:"docker-endpoint-id,omitempty"`

        // Docker network ID
        DockerNetworkID string `json:"docker-network-id,omitempty"`

        // K8s pod for this endpoint
        PodName string `json:"pod-name,omitempty"`
}

type EndpointNetworking struct {

        // IP4/6 addresses assigned to this Endpoint
        Addressing []*AddressPair `json:"addressing"`

        // host addressing
        HostAddressing *NodeAddressing `json:"host-addressing,omitempty"`

        // MAC address
        HostMac string `json:"host-mac,omitempty"`

        // Index of network device
        InterfaceIndex int64 `json:"interface-index,omitempty"`

        // Name of network device
        InterfaceName string `json:"interface-name,omitempty"`

        // MAC address
        Mac string `json:"mac,omitempty"`
}

type IPAMResponse struct {

        // address
        // Required: true
        Address *AddressPair `json:"address"`

        // host addressing
        // Required: true
        HostAddressing *NodeAddressing `json:"host-addressing"`
}
```
